oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MergeRunner.java

```diff
@@ -0,0 +1,295 @@
+ *         </ol>
+ *     </li>
+ *     <li>
+ *         We use another phaser for coordination between monitoring thread (MergeRunner) and the merge task executor (MergeTask). This phaser has one phase -
+ *         <ol>
+ *             <li>Waiting for all merge tasks complete</li>
+ *         </ol>
+ *     </li>
+ * </ol>
+ */
+public class MergeRunner implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(MergeRunner.class);
+    private final Charset charset = UTF_8;
+    private final boolean compressionEnabled;
+    private final ArrayList<File> mergedFiles = Lists.newArrayList();
+    private final ArrayList<File> unmergedFiles = Lists.newArrayList();
+    private final ExecutorService executorService;
+    private final int threadPoolSize = Integer.getInteger(PROP_MERGE_THREAD_POOL_SIZE, DEFAULT_NUMBER_OF_MERGE_TASK_THREADS);
+    private final int batchMergeSize = Integer.getInteger(PROP_MERGE_TASK_BATCH_SIZE, DEFAULT_NUMBER_OF_FILES_PER_MERGE_TASK);
+    private final Comparator fileSizeComparator = new SizeFileComparator();
+
+    /**
+     * The end result file after merging all sorted files.
+     */
+    private final File sortedFile;
+
+    /**
+     * Directory where intermediate merged files will be created.
+     */
+    private final File mergeDir;
+
+    /**
+     * Comparator used for comparing node states for creating sorted files.
+     */
+    private final Comparator<NodeStateHolder> comparator;
+    private final BlockingQueue<File> sortedFiles;
+    private final ConcurrentLinkedQueue<Throwable> throwables;
+
+    /**
+     * Phaser used for coordination with the traverse/download and sort tasks. Advance of this phaser indicates that a single
+     * merged and sorted file has been created.
+     */
+    private final Phaser phaser;
+
+    /**
+     * This poison pill is added to {@link #sortedFiles} to indicate that download phase has completed.
+     */
+    public static final File MERGE_POISON_PILL = new File("");
+
+    /**
+     * Constructor.
+     * @param sortedFiles thread safe list containing files to be merged.
+     * @param comparator comparator used to help with sorting of node state entries.
+     * @param mergeDir directory where sorted files will be created.
+     * @param compressionEnabled if true, the created files would be compressed
+     */
+    MergeRunner(File sortedFile, BlockingQueue<File> sortedFiles, File mergeDir, Comparator comparator,
+                Phaser phaser, boolean compressionEnabled) throws IOException {
+        this.mergeDir = mergeDir;
+        FileUtils.forceMkdir(mergeDir);
+        this.compressionEnabled = compressionEnabled;
+        this.sortedFiles = sortedFiles;
+        this.sortedFile = sortedFile;
+        this.throwables = new ConcurrentLinkedQueue<>();
+        this.comparator = comparator;
+        this.phaser = phaser;
+        this.executorService = Executors.newFixedThreadPool(threadPoolSize);
```

Comment:
Could you defer creating the executorService to the merge part as well? Two reasons:

(A) If only the constructor is called, but there is some exception, then we have a dangling executorService.

(B) That would allow having setters for threadPoolSize and batchMergeSize, which can then simplify testing (no need to change the system property).

Reply:
mmake sense

Reply:
I actually make it a parameter. don't think we would ever change the value, at least threadpool size cannot be changed.